src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.cs

```diff
@@ -87,6 +87,58 @@ internal static int GetCount<T>(ref IEnumerable<T> sequence)
         }
 
         /// <summary>
+        /// Tries to copy the elements in the sequence to the specified array,
+        /// if the sequence is a well-known collection type. Otherwise, does
+        /// nothing and returns <c>false</c>.
+        /// </summary>
+        /// <typeparam name="T">The type of element in the sequence.</typeparam>
+        /// <param name="sequence">The sequence to copy.</param>
+        /// <param name="array">The array to copy the elements to.</param>
+        /// <param name="arrayIndex">The index in the array to start copying.</param>
+        /// <returns><c>true</c> if the elements were successfully copied; <c>false</c> otherwise.</returns>
+        /// <remarks>
+        /// <para>
+        /// The reason we don't copy anything other than for well-known types is that a malicious interface
+        /// implementation of <see cref="ICollection{T}"/> could hold on to the array when its <see cref="ICollection{T}.CopyTo"/>
+        /// method is called. If the array it holds onto underlies an <see cref="ImmutableArray{T}"/>, it could violate
+        /// immutability by modifying the array.
+        /// </para>
+        /// </remarks>
+        internal static bool TryCopyTo<T>(this IEnumerable<T> sequence, T[] array, int arrayIndex)
+        {
+            // IList is the GCD of what the following 3 types implement.
+            var listInterface = sequence as IList<T>;
+            if (listInterface != null)
+            {
+                var sourceArray = sequence as T[];
+                if (sourceArray != null)
+                {
+                    // Note: This can have issues if T is a value type, but the underlying type
+                    // of the T[] is not typeof(T[]).
+                    // More discussion here: https://github.com/dotnet/corefx/issues/2241
```

Comment:
@AArnott Actually, it's a breaking change. It causes this app to now fail with an `ArrayTypeMismatchException`:

```cs
class Program
{
    static void Main(string[] args)
    {
        int[] ints = new[] { 1, 2, 3 };
        uint[] uints = (uint[])(object)ints;

        var immutableUints = ImmutableArray.Create(new[] { 1u, 2u, 3u });
        immutableUints.AddRange(uints);
    }
}
```

However, considering that `ImmutableArray` doesn't seem to support these types of arrays in the first place, e.g. `ImmutableArray.Create((uint[])(object)new[] { 1, 2, 3 })` likewise throws an `ArrayTypeMismatchException` today, and given that these kinds of arrays are quite rare, it seems reasonable to make such a change.